docs/src/components/home/ElementPointer.tsx

```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import Box, { BoxProps } from '@material-ui/core/Box';
+import Typography from '@material-ui/core/Typography';
+import { debounce } from '@material-ui/core/utils';
+
+const PointerContext = React.createContext<undefined | React.Dispatch<React.SetStateAction<Data>>>(
+  undefined,
+);
+
+export const withPointer = <T extends React.ElementType>(
+  Component: T,
+  options: { id: string; name: string },
+) => {
+  const WithPointer = (props: unknown) => {
+    const root = React.useRef<null | HTMLElement>(null);
+    const setData = React.useContext(PointerContext);
+    return (
+      <React.Fragment>
+        {/* @ts-ignore */}
+        <Component
+          ref={root}
+          {...props}
+          onMouseOver={(event: React.MouseEvent) => {
+            event.stopPropagation();
+            if (setData && root.current) {
+              setData({
+                id: options.id,
+                target: root.current,
+                name: options.name,
+              });
+            }
+          }}
+        />
+      </React.Fragment>
+    );
+  };
+
+  return WithPointer as T;
+};
+
+export type Data = { id: null | string; name: null | string; target: null | HTMLElement };
+
+export default function PointerContainer({
```

Comment:
Could you explain what kind of UI you're implementing here? The used patterns are new to me and have some quirks with concurrent rendering since they're not using strict React.